File: client/src/discord_bridge/commands.py
"""Discord command handling.

Parses /agentname message and routes to relay.
Handles file uploads and unknown user detection.
"""

from __future__ import annotations

import os
from pathlib import Path

import discord
import structlog

from ..config import Config
from ..relay.client import RelayClient
from .onboarding import OnboardingHandler
from .formatter import format_confirmation

log = structlog.get_logger()


class CommandHandler:
    """Parses Discord commands and forwards to relay."""

    def __init__(self, config: Config, relay: RelayClient):
        self.config = config
        self.relay = relay
        self.onboarding = OnboardingHandler(config, relay)
        self._agent_names: set[str] | None = None

    async def _get_agent_names(self) -> set[str]:
        """Cache known agent names from relay."""
        if self._agent_names is None:
            try:
                agents = await self.relay.list_agents()
                self._agent_names = {a.name for a in agents}
            except Exception:
                self._agent_names = set(self.config.agents.keys())
        return self._agent_names

    def _resolve_sender(self, author: discord.Member | discord.User) -> str | None:
        """Map Discord user to relay profile slug."""
        discord_id = str(author.id)
        for slug, human in self.config.humans.items():
            if human.discord_id == discord_id:
                return slug
        return None

    async def handle(self, message: discord.Message):
        """Handle a /command message."""
        content = message.content.lstrip("/")
        if not content:
            return

        parts = content.split(None, 1)
        target = parts[0].lower()
        body = parts[1] if len(parts) > 1 else ""

        await self.handle_raw(message, target, body)

    async def handle_raw(self, message: discord.Message, target: str, body: str):
        """Handle a parsed command: target agent + message body."""
        # Resolve sender
        sender = self._resolve_sender(message.author)
        if sender is None:
            # Unknown user → trigger onboarding
            await self.onboarding.handle_unknown_user(message)
            return

        # Check sender has access to target's pool
        if not self._check_pool_access(sender, target):
            await message.reply(
                f"You don't have access to talk to **{target}**.",
                mention_author=False,
            )
            return

        # Validate target agent exists
        agent_names = await self._get_agent_names()
        if target not in agent_names:
            available = ", ".join(sorted(agent_names))
            await message.reply(
                f"Agent **{target}** not found. Available: {available}",
                mention_author=False,
            )
            return

        # Handle file attachments
        file_paths = await self._download_attachments(message)
        if file_paths:
            file_info = "\n".join(f"[Attached: {p}]" for p in file_paths)
            body = f"{body}\n\n{file_info}" if body else file_info

        if not body:
            await message.reply(
                f"Usage: `/{target} your message here`",
                mention_author=False,
            )
            return

        # Determine message type
        msg_type = "question" if body.rstrip().endswith("?") else "notification"

        # Derive subject from first ~8 words
        words = body.split()[:8]
        subject = " ".join(words)
        if len(body.split()) > 8:
            subject += "..."

        # Send to relay
        try:
            await self.relay.send_message(
                as_=sender,
                to=target,
                subject=subject,
                content=body,
                type=msg_type,
            )
            # Confirm with a reaction
            await message.add_reaction("\u2705")
            log.info("discord.forwarded", sender=sender, target=target, chars=len(body))
        except Exception as e:
            log.error("discord.forward_failed", error=str(e))
            await message.reply(
                f"Failed to send message: {e}",
                mention_author=False,
            )

    def _check_pool_access(self, sender_slug: str, target_agent: str) -> bool:
        """Check if sender has access to the target agent's pool."""
        human = self.config.humans.get(sender_slug)
        if not human:
            return False

        # Find which pool the target belongs to
        target_pool = None
        agent_cfg = self.config.agents.get(target_agent)
        if agent_cfg:
            target_pool = agent_cfg.pool

        if target_pool is None:
            # Agent not in config, allow (might be dynamically registered)
            return True

        return target_pool in human.pools

    async def _download_attachments(self, message: discord.Message) -> list[str]:
        """Download .txt and .md attachments to local disk."""
        paths = []
        download_dir = Path(self.config.machine.download_dir)
        download_dir.mkdir(parents=True, exist_ok=True)

        for attachment in message.attachments:
            if not attachment.filename.endswith((".txt", ".md")):
                continue

            dest = download_dir / attachment.filename
            # Avoid overwriting
            if dest.exists():
                stem = dest.stem
                suffix = dest.suffix
                counter = 1
                while dest.exists():
                    dest = download_dir / f"{stem}_{counter}{suffix}"
                    counter += 1

            try:
                await attachment.save(dest)
                paths.append(str(dest))
                log.info("discord.file_downloaded", file=str(dest), size=attachment.size)
            except Exception as e:
                log.error("discord.download_failed", file=attachment.filename, error=str(e))

        return paths

File: internal/discord/commands.go
package discord

import (
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"sync"

	"github.com/bwmarrin/discordgo"

	"agent-relay/internal/client"
)

// CommandHandler parses /agentname commands from Discord and routes to relay.
type CommandHandler struct {
	config     *client.Config
	relay      *client.RelayClient
	onboarding *OnboardingHandler

	agentNames   map[string]bool
	agentNamesMu sync.RWMutex
}

// NewCommandHandler creates a command handler.
func NewCommandHandler(cfg *client.Config, relay *client.RelayClient) *CommandHandler {
	return &CommandHandler{
		config:     cfg,
		relay:      relay,
		onboarding: NewOnboardingHandler(cfg, relay),
	}
}

// Handle processes a /command message from Discord.
func (c *CommandHandler) Handle(s *discordgo.Session, m *discordgo.MessageCreate) {
	content := strings.TrimPrefix(m.Content, "/")
	if content == "" {
		return
	}

	parts := strings.SplitN(content, " ", 2)
	target := strings.ToLower(parts[0])
	body := ""
	if len(parts) > 1 {
		body = parts[1]
	}

	// Resolve sender
	sender := c.resolveSender(m.Author.ID)
	if sender == "" {
		c.onboarding.HandleUnknownUser(s, m)
		return
	}

	// Validate team access via relay teams (G6)
	if !c.checkAccess(sender, target) {
		s.ChannelMessageSend(m.ChannelID,
			fmt.Sprintf("You don't have access to talk to **%s**.", target))
		return
	}

	// Validate target exists
	if !c.isValidAgent(target) {
		names := c.getAgentNames()
		s.ChannelMessageSend(m.ChannelID,
			fmt.Sprintf("Agent **%s** not found. Available: %s", target, strings.Join(names, ", ")))
		return
	}

	// Handle file attachments
	filePaths := c.downloadAttachments(m)
	if len(filePaths) > 0 {
		fileInfo := make([]string, len(filePaths))
		for i, p := range filePaths {
			fileInfo[i] = fmt.Sprintf("[Attached: %s]", p)
		}
		if body != "" {
			body += "\n\n" + strings.Join(fileInfo, "\n")
		} else {
			body = strings.Join(fileInfo, "\n")
		}
	}

	if body == "" {
		s.ChannelMessageSend(m.ChannelID,
			fmt.Sprintf("Usage: `/%s your message here`", target))
		return
	}

	// Determine message type and subject
	msgType := "notification"
	if strings.HasSuffix(strings.TrimSpace(body), "?") {
		msgType = "question"
	}

	words := strings.Fields(body)
	subject := strings.Join(words, " ")
	if len(words) > 8 {
		subject = strings.Join(words[:8], " ") + "..."
	}

	// Send to relay
	if err := c.relay.SendMessage(sender, target, subject, body, msgType, ""); err != nil {
		log.Printf("[discord] forward failed: %v", err)
		s.ChannelMessageSend(m.ChannelID,
			fmt.Sprintf("Failed to send message: %v", err))
		return
	}

	// Confirm with reaction
	s.MessageReactionAdd(m.ChannelID, m.ID, "✅")
	log.Printf("[discord] forwarded: %s → %s (%d chars)", sender, target, len(body))
}

func (c *CommandHandler) resolveSender(discordID string) string {
	for slug, human := range c.config.Humans {
		if human.DiscordID == discordID {
			return slug
		}
	}
	return ""
}

// checkAccess validates that the sender can talk to the target.
// Uses relay teams (G6) — not just client-side pool config.
func (c *CommandHandler) checkAccess(senderSlug, targetAgent string) bool {
	human, ok := c.config.Humans[senderSlug]
	if !ok {
		return false
	}

	// Find target agent's pool
	agentCfg, ok := c.config.Agents[targetAgent]
	if !ok {
		// Agent not in config — allow (might be dynamically registered)
		return true
	}

	// Check if sender has access to target's pool
	for _, pool := range human.Pools {
		if pool == agentCfg.Pool {
			return true
		}
	}

	return false
}

func (c *CommandHandler) isValidAgent(name string) bool {
	// Check config first
	if _, ok := c.config.Agents[name]; ok {
		return true
	}

	// Check cached relay agents
	c.agentNamesMu.RLock()
	if c.agentNames != nil {
		_, ok := c.agentNames[name]
		c.agentNamesMu.RUnlock()
		return ok
	}
	c.agentNamesMu.RUnlock()

	// Fetch from relay
	c.refreshAgentNames()

	c.agentNamesMu.RLock()
	defer c.agentNamesMu.RUnlock()
	if c.agentNames != nil {
		_, ok := c.agentNames[name]
		return ok
	}
	return false
}

func (c *CommandHandler) refreshAgentNames() {
	data, err := c.relay.ListAgents()
	if err != nil {
		return
	}

	var resp struct {
		Agents []struct {
			Name string `json:"name"`
		} `json:"agents"`
	}
	if err := json.Unmarshal(data, &resp); err != nil {
		return
	}

	names := make(map[string]bool)
	for _, a := range resp.Agents {
		names[a.Name] = true
	}

	c.agentNamesMu.Lock()
	c.agentNames = names
	c.agentNamesMu.Unlock()
}

func (c *CommandHandler) getAgentNames() []string {
	c.refreshAgentNames()

	c.agentNamesMu.RLock()
	defer c.agentNamesMu.RUnlock()

	var names []string
	for n := range c.agentNames {
		names = append(names, n)
	}
	return names
}

func (c *CommandHandler) downloadAttachments(m *discordgo.MessageCreate) []string {
	var paths []string
	downloadDir := c.config.Machine.DownloadDir
	os.MkdirAll(downloadDir, 0755)

	for _, att := range m.Attachments {
		if !strings.HasSuffix(att.Filename, ".txt") && !strings.HasSuffix(att.Filename, ".md") {
			continue
		}

		dest := filepath.Join(downloadDir, att.Filename)
		// Avoid overwriting
		for i := 1; fileExists(dest); i++ {
			ext := filepath.Ext(att.Filename)
			name := strings.TrimSuffix(att.Filename, ext)
			dest = filepath.Join(downloadDir, fmt.Sprintf("%s_%d%s", name, i, ext))
		}

		resp, err := http.Get(att.URL)
		if err != nil {
			log.Printf("[discord] download failed: %v", err)
			continue
		}

		data, err := io.ReadAll(resp.Body)
		resp.Body.Close()
		if err != nil {
			continue
		}

		if err := os.WriteFile(dest, data, 0644); err != nil {
			continue
		}

		paths = append(paths, dest)
		log.Printf("[discord] downloaded: %s (%d bytes)", dest, len(data))
	}
	return paths
}

func fileExists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}
